File: engine/world.lua
local World = {}
World.__index = World

--
-- create a new world map
-- 
function World:new(x1, y1, x2, y2)
   local self = setmetatable({}, World)

   self.x1, self.y1, self.x2, self.y2 = x1, y1, x2, y2

   self.objects = {}
   self.cities = {}
   self.predefined_tiles = {}

   self.__obj_ptr = {}
   self.__tile_cache = {}

   self.objects_ptr = {}
   return self
end


-- 
-- initialize world map
--
function World:initialize()
   -- create map
   self.mapgen = MapGen:new(self.x1, self.y1, self.x2, self.y2)
   self.mapgen:create()

   -- create cities
   self:__create_cities()
   self.mapgen:create_roads()

   self:__init_physics()

   self.player = self:__add_object(Player:new(5, 10))
   local players_vehicle = self:__add_object(Vehicle:new(10, 10, VehicleModel.REGULAR))
   self.player.vehicle = players_vehicle
   
   self:__add_people_to_cities()

   self:__add_static_objects()
   collectgarbage()
end


--
-- clean up
--
function World:clean_up()
   for _,obj in ipairs(self.objects) do
      self:__remove_object(obj)
   end
   self:__finish_physics()
end


--
-- one step in the world
--
function World:step()
   self:__physics_step()
   for _,object in ipairs(self.objects) do
      object:step()
   end
end


--
-- return the stack of tiles (max 10)
--
function World:tiles(x, y)
   local predef = self.predefined_tiles[world:unique_tile_id(x,y)]
   if predef then
      return predef
   else
      for _, city in ipairs(self.cities) do
         if x >= city.x and x < (city.x+city.w) and y >= city.y and y < (city.y+city.h) then
            return city:tiles(x-city.x, y-city.y)
         end
      end
      return { self.mapgen:tile(x,y) } -- TODO
   end
end


--
-- return the stack of tile IDs (max 10)
--
function World:tilemap_ids(x1, y1, x2, y2)
   -- TODO - cache size
   local t = {}
   local __tile_cache = self.__tile_cache -- speedup hack
   for x = x1,x2 do
      if not __tile_cache[x] then __tile_cache[x] = {} end
      t[x] = {}
      for y = y1,y2 do
         local ids = __tile_cache[x][y]
         if not ids then
            ids = self:tiles(x,y)
            for k,v in ipairs(ids) do
               ids[k] = v.id
            end
            __tile_cache[x][y] = ids
         end
         t[x][y] = ids
      end
   end
   self.__tile_cache = __tile_cache
   return t
end


-- 
-- return a list of objects among the tiles
--
function World:objects_in_area(x1, y1, x2, y2)
   local cond = function(p) 
      local pos = p:pos()
      return (pos.x >= x1 and pos.x <= x2 and pos.y >= y1 and pos.y <= y2)
   end
   return funct.filter(self.objects, cond)
end


-- 
-- return a unique identifier for a tile, and revert the value
--
function World:unique_tile_id(x, y)
   return x + (y * 1000000)
end
function World:revert_unique_tile(id)
   return (id % 1000000), math.floor(id / World.W)
end


--
-- return if this tile can be walked by a person
--
function World:tile_walkable(x, y)
   local st = self:tiles(x, y)
   if st[1].friction == math.huge or (st[2] and st[2].solid) or (st[3] and st[3].solid) then
      return false
   end
   return true
end


--
-- when a collision happens, this method is called by the C code
--
function World:collision_callback(body_a, body_b)
   local a = self.__obj_ptr[body_a]
   local b = self.__obj_ptr[body_b]
   a:collision(b)
   b:collision(a)
end


-- 
-- Return the limits of the world (x1, y1, x2, y2)
-- 
function World:limits()
   return self.x1, self.y1, self.x2, self.y2
end


-- 
-- Return type
--
function World:type()
   return 'World'
end


--
-- Print some debug information
--
function World:debug()
   print 'something'
end


-------------
-- PRIVATE --
-------------

function World:__tostring()
   return '[World]'
end


function World:__create_cities()
   for _,point in ipairs(self.mapgen:cities_positions(20)) do
      local r, tp = math.random(), nil
      if r < 0.1 then
         tp = CityType.CAPITAL
      elseif r < 0.2 then
         tp = CityType.FRONTIER
      else
         tp = self:__city_type(point)
      end
      -- TODO - random
      self.cities[#self.cities+1] = City:new(point, tp)
      --self.cities[#self.cities+1] = City:new(CityLayout.LAYOUT_1, p.x, p.y, 20, 20, Block.GRASS)
   end
end


function World:__city_type(p)
   local b = self.mapgen.plane:polygon_containing_point(p)
   if b == Biome.GRASS then
      return CityType.AGRICULTURAL
   elseif b == Biome.SNOW then
      return CityType.REFINERY
   elseif b == Biome.BARE then
      return CityType.MINING
   elseif b == Biome.TEMPFOR or b == Biome.TROPFOR then
      return CityType.FORESTAL
   elseif b == Biome.DESERT or b == Biome.BEACH then
      return CityType.CHEMICAL
   end
   return CityType.RANDOM
end


function World:__add_people_to_cities()
   for _,city in ipairs(self.cities) do
      for _,building in ipairs(city.buildings) do
         if building.layout.people then
            for _,p in ipairs(building.layout.people) do
               -- create person
               local person = nil
               local x = p.x + city.x + building.x + 0.5
               local y = p.y + city.y + building.y + 0.5
               if p.type == 'Shopkeeper' then
                  person = ai.Shopkeeper:new(x, y) -- TODO (?)
                  self:__add_object(person)
               end
               assert(person, 'Invalid person type: ' .. p.type)
            end
         end
      end
   end
end


function World:__add_object(obj)
   self.objects[#self.objects+1] = obj
   obj:init_physics()
   self.__obj_ptr[obj.body] = obj
   return obj
end


function World:__remove_object(obj)
   obj:clean_up()
   for k,v in ipairs(obj) do 
      if v == obj then table.remove(self.objects, k) end
   end
end


-- TODO : add static objects outside cities
function World:__add_static_objects()
   for _,city in ipairs(self.cities) do
      for x = city.x, (city.x+city.w) do
         for y = city.y, (city.y+city.h)  do
            if not self:tile_walkable(x,y) then
               self:__physics_create_static_obj(x, y, 1, 1)
            end
         end
      end
   end
end

return World

-- vim: ts=3:sw=3:sts=3:expandtab

File: engine/world.cc
#include <chipmunk.h>

#include <cstdlib>
#include <iostream>
using namespace std;

#ifdef __MINGW32__
#  define rand_r(x) rand()
#endif

#include "engine/world.h"

#include "./main.h"
#include "engine/city.h"
#include "engine/citylayout.h"
#include "engine/hero.h"
#include "engine/vehicle.h"
#include "engine/mapgen/mapgen.h"

World::World(int x1, int y1, int x2, int y2, unsigned int seedp)
    : Blocks(), x1(x1), y1(y1), x2(x2), y2(y2), seedp(seedp),
      mapgen(MapGen(x1, y1, x2, y2))
{ 
    cpSpaceSetDefaultCollisionHandler(space, CollisionCallback, 
            NULL, NULL, NULL, this);
}


World::~World()
{
    // cleanup physics
    cpBodyEachShape(space->staticBody, FreeStaticShape, this);
}

void
World::Initialize()
{
    // create map
    mapgen.Create();

    // initialize cities
    CreateCities();
    City& newhope = cities[0];

    // initialize vehicles
    shared_ptr<Vehicle> car(new Vehicle(Point(newhope.X+10, newhope.Y+10), VehicleModel::GENERIC));
    AddObject(car);

    // initialize hero
    hero = shared_ptr<class Hero>(new class Hero(Point(newhope.X, newhope.Y), car));
    AddObject(hero);

    // create roads
    mapgen.CreateRoads(cities);

    // create people
    AddWorkers();
    AddStaticObjects();
}


void
World::Step()
{
    cpSpaceStep(space, 1.0/60.0);
    for(auto& obj : objects) {
        obj->Step();
    }
}


int
World::Tiles(Block (&block)[10], int x, int y) const
{
    // TODO - check from cache

    for(auto const& city: cities) {
        if(x >= city.X && x<(city.X+city.W())
        && y >= city.Y && y<(city.Y+city.H())) {
            int n = city.Tiles(block, x-city.X, y-city.Y);
            if(block[0] == Block::EMPTY) {
                block[0] = mapgen.Terrain(x, y);
            }
            return n;
        }
    }

    block[0] = mapgen.Terrain(x, y);
    return 1;
}


bool 
World::IsTileWalkable(int x, int y) const
{
    Block block[10];
    int n = Tiles(block, x, y);

    if(n == 2 && !Blocks.Examine(block[1]).Crossable) {
        return false;
    }
    if(n >= 3 && !Blocks.Examine(block[1]).Crossable) {
        return false;
    }

    return true;
}


void 
World::Limits(int& x1, int& y1, int& x2, int& y2) const
{
    x1 = this->x1;
    y1 = this->y1;
    x2 = this->x2;
    y2 = this->y2;
}


void 
World::RecalculatePrices()
{
    // TODO
}


/*****************************************************************************/


double 
World::Random() const
{
    return (static_cast<double>(rand_r(&seedp)) / static_cast<double>(RAND_MAX));
}


void
World::CreateCities()
{
    unordered_set<Point> positions = mapgen.CitiesPositions(30);
    for(auto const& pos : positions) {
        int x = static_cast<int>(pos.X()),
            y = static_cast<int>(pos.Y());
        double r = Random();
        CityType type;
        if(r < 0.1) {
            type = CityType::CAPITAL;
        } else if(r < 0.2) {
            type = CityType::FRONTIER;
        } else {
            type = City::Type(mapgen.Terrain(x, y));
        }
        cities.emplace_back(x, y, type, 1);
    }

    cities.emplace_back(0, 0, CityType::AGRICULTURAL, 1); // TODO
}


void 
World::AddObject(Object::Ptr obj)
{
    objects.push_back(obj);
    physics_ptr[obj->PhysicsBodyPtr()] = obj;
}


void
World::AddWorkers()
{
    for(auto& city: cities) {
        for(auto const& worker : city.Workers()) {
            shared_ptr<Worker> wk = Worker::MakeWorker(worker.second, city, worker.first);
            AddObject(wk);
        }
    }
}


void
World::AddStaticObjects()
{
    for(auto const& city : cities) {
        for(int x=city.X; x<city.X+city.W(); x++) {
            for(int y=city.Y; y<city.Y+city.H(); y++) {
                if(!IsTileWalkable(x, y)) {
                    AddStaticObject(x+0.5, y+0.5, 1, 1);
                }
            }
        }
    }
}


void 
World::AddStaticObject(double x, double y, double w, double h)
{
    cpShape *shape = cpBoxShapeNew2(space->staticBody, 
            cpBBNew(x-(w/2), y-(h/2), x+(w/2), y+(h/2)));
    cpShapeSetFriction(shape, 1);
    cpSpaceAddShape(space, shape);
}


void 
World::FreeStaticShape(cpBody *body, cpShape *shape, void* data)
{
    (void) body; (void) data;

    cpSpaceRemoveShape(space, shape);
    cpShapeFree(shape);
}


int 
World::CollisionCallback(struct cpArbiter *arb, struct cpSpace *space, 
        void *data)
{
    World* world = static_cast<World*>(data);

    cpBody *a, *b;
    cpArbiterGetBodies(arb, &a, &b);

    // ingnore static objects (TODO)
    if(a == space->staticBody || b == space->staticBody)
        return 1;

    // call both objects 'collision'
    world->physics_ptr[a]->Collision(world->physics_ptr[b]);
    world->physics_ptr[b]->Collision(world->physics_ptr[a]);

    return 1;
}


// vim: ts=4:sw=4:sts=4:expandtab
